bndtools.m2e/src/bndtools/m2e/MavenWorkspaceRepository.java

```diff
@@ -0,0 +1,336 @@
+    private boolean inited = false;
+
+    private final Map<String,Entry<IMavenProjectFacade,MavenProject>> bsnMap = new HashMap<>();
+
+    @Override
+    public Map<Requirement,Collection<Capability>> findProviders(Collection< ? extends Requirement> requirements) {
+        return null;
+    }
+
+    @Override
+    public boolean refresh() throws Exception {
+        return false;
+    }
+
+    @Override
+    public File getRoot() throws Exception {
+        return null;
+    }
+
+    @Override
+    public PutResult put(InputStream stream, PutOptions options) throws Exception {
+        return null;
+    }
+
+    @Override
+    public File get(final String bsn, final Version version, Map<String,String> properties, final DownloadListener... listeners) throws Exception {
+        if (!inited) {
+            init();
+        }
+
+        final Entry<IMavenProjectFacade,MavenProject> entry = bsnMap.get(bsn);
+
+        if (entry == null)
+            return null;
+
+        final MavenProject mavenProject = entry.getValue();
+
+        // add the eclipse project that this comes from so we can look it up in the launch
+        // see bndtools.launch.BndDependencySourceContainer.createSourceContainers()
+        final String projectName = entry.getKey().getProject().getName();
+        properties.put("sourceProjectName", projectName);
+
+        final File bundleFile = new File(mavenProject.getBuild().getDirectory(), mavenProject.getBuild().getFinalName() + ".jar");
+
+        if (listeners == null || listeners.length == 0) {
+            String errorIfAny = createBundleFileIfNeeded(bundleFile, bsn, version, mavenProject, new NullProgressMonitor());
+
+            if (errorIfAny != null) {
+                return null;
+            }
+
+            return bundleFile;
+        }
+
+        new Job(getName() + " getting " + bsn) {
+            @Override
+            public IStatus run(IProgressMonitor monitor) {
+                String errorIfAny = createBundleFileIfNeeded(bundleFile, bsn, version, mavenProject, monitor);
+
+                for (DownloadListener listener : listeners) {
+                    try {
+                        if (errorIfAny == null) {
+                            listener.success(bundleFile);
+                        } else {
+                            listener.failure(bundleFile, errorIfAny);
+                        }
+                    } catch (Throwable t) {
+                        logger.logError("Download listener error", t);
+                    }
+                }
+
+                return Status.OK_STATUS;
+            }
+        }.schedule();
+
+        return bundleFile;
+    }
+
+    private String createBundleFileIfNeeded(File bundleFile, String bsn, Version version, MavenProject mavenProject, IProgressMonitor monitor) {
+        String failureReason = null;
+
+        if (isBundleFileOutOfDate(bundleFile, mavenProject)) {
+            try {
+                MavenUtil.execJarMojo(mavenProject, monitor);
```

Comment:
I am not clear how the exec jar mojo in here (Maven workspace repo) is related to the exec jar mojo in the BndConfigurator class. So there are two places where we trigger the jar mojo. Are they complementary? Duplicative? Perhaps this code should trigger the build in BndConfigurator instead of directly calling execJarMojo here?

Perhaps doc comments to explain the relationship?